File: js/findpeaks.js
// Dodgy O(n) peak finding algorithm written by me
function findPeaks(arr) {
	let last = arr[0];
	let lastBigger = true;
	let peaks = [];
	for (let i = 1; i < arr.length; i++) {
		if (arr[i] <= last && lastBigger) {
			peaks.push(last); // Peak index is i - 1
		}
		lastBigger = arr[i] >= last;
		last = arr[i];
	}
	return peaks;
}

console.log(findPeaks([8, 6, 4, 9, 7, 7, 7, 3, 10, 5])); // Expected output: [8, 9, 7, 7, 10]

File: cpp/findpeaks.cpp
#include <iostream>

// My O(n) peak finding algorithm
template <typename T> void findPeaks(T arr[], int size)
{
	T prev = arr[0];
	bool prevBigger = true;
	for (int i = 1; i < size; i++)
	{
		if (arr[i] <= prev && prevBigger)
		{
			std::cout << "Peak at " << i - 1 << ": " << prev << '\n';
		}
		prevBigger = arr[i] >= prev;
		prev = arr[i];
	}
}

int main()
{
	int nums[] = { 8, 6, 4, 9, 7, 7, 7, 3, 10, 5 };
	findPeaks(nums, 10); // Expected output: 8, 9, 7, 7, 10
	
	return 0;
}